go-controller/pkg/ovn/egressfirewall_dns.go

```diff
@@ -0,0 +1,267 @@
+	for {
+		tm, dnsName, ok := e.GetNextQueryTime()
+		if !ok {
+			duration = 30 * time.Minute
+		} else {
+			now := time.Now()
+			if tm.After(now) {
+				duration = tm.Sub(now)
+			} else {
+				_, err := e.Update(dnsName)
+				if err != nil {
+					utilruntime.HandleError(err)
+				}
+
+				if !reflect.DeepEqual(e.dnsResolves[dnsName], e.dns.GetIPs(dnsName)) {
+					ipList := buildIPList(e.dns.GetIPs(dnsName))
+
+					if namespaces, exists := e.dnsInNamespace[dnsName]; exists {
+						for namespace := range namespaces {
+							for _, joinSwitch := range e.joinSwitches {
+								//delete possible old rule
+								if e.dnsResolves[dnsName] != nil {
+
+									match := fmt.Sprintf("\"%s && ip4.src == $%s\"", buildIPList(e.dnsResolves[dnsName]), e.acls[namespace][dnsName].hashedAddressSet)
+									if e.acls[namespace][dnsName].portMatch != "" {
+										match = fmt.Sprintf("%s && ( %s )\"", match[:len(match)-1], e.acls[namespace][dnsName].portMatch)
+
+									}
+									uuid, stderr, err := util.RunOVNNbctl("--data=bare", "--no-heading",
+										"--columns=_uuid", "find", "ACL",
+										fmt.Sprintf("external-ids:egressFirewall=%s", namespace),
+										fmt.Sprintf("match=%s", match))
+									if err != nil {
+										klog.Errorf("Error executing find ACL command, stderr: %q, %+v", stderr, err)
+									}
+									_, stderr, err = util.RunOVNNbctl("remove", "logical_switch",
+										joinSwitch, "acls", uuid)
+									if err != nil {
+										klog.Errorf("Failed to delete the rules for "+
+											"egressFirewall in namespace %s on node %s, stderr: %q (%v)", namespace, joinSwitch, stderr, err)
+									}
+								}
+								match := fmt.Sprintf("match=\"%s && ip4.src == $%s\"", ipList, e.acls[namespace][dnsName].hashedAddressSet)
+								if e.acls[namespace][dnsName].portMatch != "" {
+									match = fmt.Sprintf("%s && ( %s )\"", match[:len(match)-1], e.acls[namespace][dnsName].portMatch)
+
+								}
+								err := createACLRule(e.acls[namespace][dnsName].priority, match, e.acls[namespace][dnsName].action, namespace, joinSwitch)
+								if err != nil {
+									klog.Error(err)
+								}
+							}
+						}
+
+					}
+					e.dnsResolves[dnsName] = e.dns.GetIPs(dnsName)
+				}
+				continue
+			}
+		}
+
+		// Wait for the given duration or till something got added
+		select {
+		case <-e.added:
+		case <-time.After(duration):
+		}
+	}
+}
+
+func buildIPList(ipSlice []net.IP) string {
+	formattedIPList := fmt.Sprintf("ip4.dst == %s", ipSlice[0].String())
+	if len(ipSlice) == 1 {
+		//if the size is one no logic is needed
+		return formattedIPList
+	}
+	for _, ipAddress := range ipSlice {
+		formattedIPList = formattedIPList + " || ip4.dst == " + ipAddress.String()
+	}
+
+	return fmt.Sprintf("(%s)", formattedIPList)
+}
+
+func (e *EgressDNS) GetNextQueryTime() (time.Time, string, bool) {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+	tm, dnsName, timeSet := e.dns.GetNextQueryTime()
+	if !timeSet {
```

Comment:
@JacobTanenbaum this if block seems unecessary, the function returns the same thing for both cases

Reply:
done